### ISSUE TITLE
Contributing is hard

### ISSUE THREAD

**Issue by StarpTech** [carvel-dev/kbld#54]
I run into the following issues:

- Code can't be easily tested in an IDE because the setup of the tests isn't part of the go code.
- No easy way to debug it.
- Specific pack, ytt version is required. This can be automated.
- Running a test for a small cli change takes too much time. (~2min)
- Assertion fails are hard to read.
- No unit tests that verifying the correct work with programs like docker
- E2E tests only verify that kbld doesn't exit with an error code.

**Comment by StarpTech** [comment]
Additionally, when running

```sh
./hack/build.sh

eval $(minikube docker-env)
docker login ...
export KBLD_E2E_DOCKERHUB_USERNAME=...
./hack/test-all.sh
```
it produces
```
Running 'kbld package -f - --output /tmp/kbld-test-pkg-unpkg-successful --yes'...
Running 'kbld unpackage -f - --input /tmp/kbld-test-pkg-unpkg-successful --repository docker.io/starptech/kbld-test-pkg-unpkg --yes'...
    kbld.go:75: Failed to successfully execute 'kbld unpackage -f - --input /tmp/kbld-test-pkg-unpkg-successful --repository docker.io/starptech/kbld-test-pkg-unpkg --yes': Execution error: stdout: '' stderr: 'unpackage | importing 2 images...
        unpackage | importing gcr.io/cloud-builders/gcs-fetcher@sha256:055519529bf1ba12bf916fa42d6d3f68bdc581413621c269425bb0fee2467a93 -> index.docker.io/starptech/kbld-test-pkg-unpkg@sha256:055519529bf1ba12bf916fa42d6d3f68bdc581413621c269425bb0fee2467a93...
        unpackage | importing index.docker.io/library/redis@sha256:000339fb57e0ddf2d48d72f3341e47a8ca3b1beae9bdcb25a96323095b72a79b -> index.docker.io/starptech/kbld-test-pkg-unpkg@sha256:000339fb57e0ddf2d48d72f3341e47a8ca3b1beae9bdcb25a96323095b72a79b...
        unpackage | imported 0 images
        kbld: Error: Importing image gcr.io/cloud-builders/gcs-fetcher@sha256:055519529bf1ba12bf916fa42d6d3f68bdc581413621c269425bb0fee2467a93:
          Verifying imported image index.docker.io/starptech/kbld-test-pkg-unpkg:kbld-sha256-055519529bf1ba12bf916fa42d6d3f68bdc581413621c269425bb0fee2467a93:
            GET https://index.docker.io/v2/starptech/kbld-test-pkg-unpkg/manifests/kbld-sha256-055519529bf1ba12bf916fa42d6d3f68bdc581413621c269425bb0fee2467a93:
              UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:starptech/kbld-test-pkg-unpkg Type:repository]]
        ' error: 'exit status 1'
--- FAIL: TestPkgUnpkgSuccessful (46.
```

**Comment by cppforlife** [comment]
thanks for taking time to record your observations/sugestions. i'll try to provide more context on each bullet below...

> Code can't be easily tested in an IDE because the setup of the tests isn't part of the go code.

is it regarding env variables? is there setup step to do for IDEs to provide the necessary configuration? im not too familiar with IDEs so do tell us more about your setup. (i've been mostly using terminal for running tests hence my unfamiliarity.)

> No easy way to debug it.

could you be a bit more specific here whats being debugged? are you referring to debugging binary that gets called out during the test?

> Specific pack, ytt version is required. This can be automated.

for test setup we try to be very specific to avoid surprising breakages (for pack binary). there is probably more to do here, but i definitely like when the repo specifies dependency versions required for tests. could you be more specific at what you would want to automate (there are a few possibilities here i can think of).

dependency for ytt during build does not specify version afaik.

> Running a test for a small cli change takes too much time. (~2min)

that's a result of favoring e2e tests in majority of cases. you should also be able to just run specific tests (-run go test flag) that relate to your change and run the entire test suite at the end to check for any regressions.

see general testing note at the bottom as well.

> Assertion fails are hard to read.

i've tried to stick with vanilla go testing after using various other testing libraries in past years. it has been an interesting adventure in trying out something new. i dont personally feel the need to introduce something heavier at this time. if you link to some concrete assertion examples you find hard to read, we might be able to figure out how to improve them.

> No unit tests that verifying the correct work with programs like docker

see general testing note at the bottom.

> E2E tests only verify that kbld doesn't exit with an error code.

most of the tests check on kbld output, which is ultimately what we care about (example: https://github.com/k14s/kbld/blob/develop/test/e2e/resolve_test.go#L31-L43). there is probably a few places that could benefit from adding more assertions but this goes to my following point...

> ...UNAUTHORIZED: authentication required

seems like you are not logged in to docker? (https://github.com/k14s/kbld/blob/develop/docs/auth.md#via-docker-config)

---
#### testing methodology

i can summarize my last decade of using various testing approaches and methodologies in one specific insight -- tests are there to address risk of breakage. taking that point of view, hopefully it makes sense why there are unit tests for some things (e.g. https://github.com/k14s/kbld/blob/develop/pkg/kbld/search/image_refs_test.go) and not others -- because when code was written i deemed some code as more risky. beauty of this is that it's not a static determination -- over time some code will evolve to become riskier and some will derisk, so new tests will be added and some removed. of course various folks will also have different opinions of the riskiness level as well so they can contribute appropriately. another good indicator for risk level is number of bugs over time in a particular area so that's something that would influence decision to add more tests. on top of all that, by focusing more on higher level cross-component tests (one example is e2e tests) we are providing enough breathing room for implementation to change without having to change all your tests.

**Comment by StarpTech** [comment]
Thanks for the detailed response.

> is it regarding env variables? is there setup step to do for IDEs to provide the necessary configuration? im not too familiar with IDEs so do tell us more about your setup. (i've been mostly using terminal for running tests hence my unfamiliarity.)

There is no builtin support to attach a debugger to the spawned `kbld` process.

> could you be a bit more specific here whats being debugged? are you referring to debugging binary that gets called out during the test?

Exactly!

> for test setup we try to be very specific to avoid surprising breakages (for pack binary). there is probably more to do here, but i definitely like when the repo specifies dependency versions required for tests. could you be more specific at what you would want to automate (there are a few possibilities here i can think of).

+1, the installation process can be automated by a build system like make or https://github.com/go-task/task (I'm windows user)

> that's a result of favoring e2e tests in majority of cases. you should also be able to just run specific tests (-run go test flag) that relate to your change and run the entire test suite at the end to check for any regressions.

It just feels super heavy.

> i've tried to stick with vanilla go testing after using various other testing libraries in past years. it has been an interesting adventure in trying out something new. I don't personally feel the need to introduce something heavier at this time. if you link to some concrete assertion examples you find hard to read, we might be able to figure out how to improve them.

This is due to the reason that no stack can be generated when spawning the `kbld` process. Additonally, assertions are hard to read. I like https://github.com/tj/assert e.g

```
❯ go test ./...
?       ghconfig        [no test files]
ok      ghconfig/cmd    (cached)
?       ghconfig/internal/common        [no test files]
?       ghconfig/internal/config        [no test files]
--- FAIL: TestSync_MergeDependabot (0.00s)
    merge_dependabot_test.go:89:
                Error Trace:    merge_dependabot_test.go:89
                Error:          Not equal:
                                expected: dependabot.GithubDependabot{Version:"2", Updates:[]*dependabot.Updates{(*dependabot.Updates)(0xc000069400), (*dependabot.Updates)(0xc000069540), (*dependabot.Updates)(0xc000069680)}}
                                actual  : dependabot.GithubDependabot{Version:"2", Updates:[]*dependabot.Updates{(*dependabot.Updates)(0xc0000697c0), (*dependabot.Updates)(0xc000069900), (*dependabot.Updates)(0xc000069a40)}}

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -5,3 +5,3 @@
                                    PackageEcosystem: (string) "",
                                -   Directory: (string) (len=4) "/foo",
                                +   Directory: (string) (len=5) "/foo2",
                                    Schedule: (dependabot.Schedule) {
                Test:           TestSync_MergeDependabot
                Messages:       Dst is always overriden by Src
FAIL
FAIL    ghconfig/internal/dependabot    0.158s
ok      ghconfig/internal/github        (cached)
?       ghconfig/internal/helper        [no test files]
FAIL
```

> seems like you are not logged in to docker?

This looks like an issue with WSL2

> i can summarize my last decade of using various testing approaches and methodologies in one specific insight -- tests are there to address risk of breakage. taking that point of view, hopefully it makes sense why there are unit tests for some things (e.g. https://github.com/k14s/kbld/blob/develop/pkg/kbld/search/image_refs_test.go) and not others -- because when code was written i deemed some code as more risky. beauty of this is that it's not a static determination -- over time some code will evolve to become riskier and some will derisk, so new tests will be added and some removed. of course various folks will also have different opinions of the riskiness level as well so they can contribute appropriately. another good indicator for risk level is number of bugs over time in a particular area so that's something that would influence decision to add more tests. on top of all that, by focusing more on higher level cross-component tests (one example is e2e tests) we are providing enough breathing room for implementation to change without having to change all your tests.

I prefer the same approach. Testing should not make it super hard to change your code but imagine someone forgets to handle an error. In that case, kbld will process with no error because only the artifact is validated. How can you ensure reliability? Those errors can be caught before users will run into it.

In summary: I agree with the approach but some places just missing more unit or specific E2E tests (e.g check that image/tags have been really created on the registry).

Btw: A linter can be introduced to catch unhandled errors and more.

**Comment by StarpTech** [comment]
## Testing

I could run the following setup locally. You have to replace `172.17.0.0/16` with your subnetmask of `minikube ip`

```sh
minikube start --driver=docker --insecure-registry=172.17.0.0/16
./hack/build.sh
eval $(minikube docker-env)
./hack/test-all-minikube-local-registry.sh
```

Run a single test

```sh
./hack/test-all-minikube-local-registry.sh -run TestDockerBuildSuccessful
```

This should be part of the contributing.md

**Comment by cppforlife** [comment]
@StarpTech oh i guess you went the route of running local registry for testing. using dockerhub might be easier since it avoid all kinds of networking complexities dealing with minikube etc introduces.

> This should be part of the contributing.md

agreed, looks like good notes. do you mind making a PR for this?

**Comment by StarpTech** [comment]
> agreed, looks like good notes. do you mind making a PR for this?

Yes. What your thoughts about my previous points?

**Comment by StarpTech** [comment]
> using dockerhub might be easier since it avoid all kinds of networking complexities dealing with minikube etc introduces.

But it has several other drawbacks:

- Requires a docker account
- Pollutes your repositories (Free account has only 3)
- Requires good internet connection

**Comment by cppforlife** [comment]
> Yes. What your thoughts about my previous points?

i think your points about debugging and installation are fair. we have discussed some of the in the community meeting; however, we haven't settled on a concrete action plan.

**Comment by StarpTech** [comment]
Thanks for the update.

**Comment by StevenLocke** [comment]
While working on testing a PR, I ran into issues running our tests as well. `insufficient_scope: authorization failed`

After lots of digging and help from @DennisDenuto we found that I had some docker environment variables set. [0]

Unsetting them allowed kbld `hack/test-all.sh` to pass again, presumably using default values.

[0] The docker vars were:
```
DOCKER_TLS_VERIFY
DOCKER_HOST
DOCKER_CERT_PATH
MINIKUBE_ACTIVE_DOCKERD
```
